### DISCUSSION TITLE
Async/Await Not Supported in Client Components Despite Documentation Claims

### DISCUSSION THREAD

**Discussion by Kayeddy** [vercel/next.js#62031] [Help] [answered]
### Link to the code that reproduces this issue:
https://github.com/Kayeddy/threadsClone-Web

### Steps to reproduce the behavior:

Clone the linked repository.

Open the repo in vs code and run 'npm i' to install necessary dependencies.

After installing necessary dependencies run 'npm run dev', which will start up the application on the default localhost:300 server.

Head to the website and sign in (No user data is collected on registration).

On the root page (Home) try to click the like button of any Thread.

The error will pop up as soon as you hit the like button.

### Current vs. Expected behavior:

**Current Behavior:** When trying to like a post by invoking the handleThreadLike async function—which calls a server action function—and attempting to update the component to reflect the latest changes using React's useState hook within a client component in Next.js, the application throws an error indicating that async/await is not supported in client components.

**Expected Behavior:** According to the latest Next.js documentation, async/await should be supported in client components when using the app router directive. The expected behavior is for the handleThreadLike function to execute without errors, allowing for the asynchronous liking of a post.

### Additional information

```js
The issue occurs both in the development and production builds of the application.

Deployment platform: The issue has been reproduced both locally and on Vercel.

Browser: The issue has been observed in multiple browsers (Chrome, Opera, Firefox).

The use of state was initially attempted but was commented out due to the same error. A useRef hook is currently being used as a workaround.

This issue seems to contradict the latest documentation regarding the app router and client components in Next.js, suggesting either a regression or a documentation oversight.
```

**Comment by icyJoseph** [comment]
Hi,

Where in the docs do they claim that async/await works on client components?  I know that when the client component doesn't useState, it kinda works, but the moment you try to do a state update it crashes out, but I don't think I've read in the React nor Next.js docs that this is supported.

**Reply by Kayeddy** [reply]
Hello Joseph, thanks for taking the time to reply.
In regards to your question, I will leave the link to the specific section of the Docs that makes this assumption, and I will also leave a screenshot as to the example code within the Docs. Furthermore, the example within the Docs also include the use of React's useState hook alongside the use of async/await within a client component.

Link: https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations

<img width="500" alt="Screenshot 2024-02-13 200537" src="https://github.com/vercel/next.js/assets/57167218/1eef7ebc-b9a9-4c81-b7e2-80f284c8ca72">

**Reply by icyJoseph** [reply]
Aha, but you are phrasing this incorrectly...

The way you are stating this is as if you expected `LinkButton` to be an async function, which is not the case.

An `onClick` callback, has forever and ever, been able to be async/await though, so you should not really have any issues if you are doing what the example shows.

I'll take a look at your repo.

**Reply by icyJoseph** [reply, marked as answer]
Ok, 

From reading your code I think I've got it.

- https://github.com/Kayeddy/threadsClone-Web/blob/master/components/shared/ThreadInteractionSection.tsx

That's a client component, which owns:

- CommentThreadDialog

Which is defined as `async` function.

Could you remove the `async` function tag from it? This one right here: https://github.com/Kayeddy/threadsClone-Web/blob/master/components/dialogs/CommentThreadDialog.tsx#L33

**Reply by Kayeddy** [reply]
That was so easy I almost cried when it worked. So let me get this straight, there cannot be an async function component inside of a client component?

**Reply by icyJoseph** [reply]
An async function component can't be owned by a client component, however a client component can be parent to a server component though. 

I am a bit short on time today, but I'd recommend reading this section of the docs: https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns

**Comment by Kayeddy** [comment]
Many thanks @icyJoseph, I appreciate the help mate. Cheers.
